### DISCUSSION TITLE
[ feature ] embed native language support

### DISCUSSION THREAD

**Discussion by std-microblock** [ASDAlexander77/TypeScriptCompiler#56] [Q&A]
It would be cool if there's a way to directly use native languages in typescript!

Take rust as an example ( as it's also using LLVM, it would probably be easier to implement )

```typescript
// main.ts
import { read_file } from "./config.rs"

function main(){
    let config = read_file('./config.ini').value;
}
```

```rust
// config.rs
use std::fs::File;
use std::io::prelude::*;
use std::io::Error;

fn read_file(path: &str) -> Result<String, Error> {
    let mut file = File::open(path)?;
    let mut contents = String::new();
    file.read_to_string(&mut contents)?;

    Ok(contents)
}
```

This would not only bring the great stdlib and other libraries of the language to this TSC but also take advantage of the language's strong points!

**Comment by ASDAlexander77** [comment]
Have a look at project ``\TypeScriptCompiler\docs\how\cmake_vulkan\`` you will see how to use native libraries with typescript native compiler.

**Comment by ASDAlexander77** [comment]
FYI: TypeScript compiles code in the same way as any C compiler. so u can reference any function in C code and any C function in TypeScript Native Compiler ... as all code will be compiled into OBJ files LINKER will resolve names.

**Comment by ASDAlexander77** [comment]
second option can be to use ''import "./dll.dll"'' and then you can provide function names with "@dllimport" attributes

for example. if you have dll with function "point"

```TypeScript
type Point = { x: number; y: number; };

import "./<any_dll_file_without_extention_name>"

@dllimport
declare function point (x: number, y: number): Point;

function main() {
    const p = point (1.0, 2.0);
    print(`x=${p.x}, y=${p.y}`);

    print("done.");
}
```
